### PR TITLE
JAMES-2776 DomainsRoutes should not accept same source & destination

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

**Comment by trantienduchn** [comment]
```
[55253ab273a9c6c7ee6f587b46ba8b82025df4ee] 2019-05-21 06:40:55,849 INFO  [main] - [🐳 [linagora/linshare-init:2.2]]- Container linagora/linshare-init:2.2 is starting: 8f7f9dcf927bfe21c71bed347dfa73a21e954976e4f2e83f6c6774bac4adf7bf
[55253ab273a9c6c7ee6f587b46ba8b82025df4ee] 2019-05-21 06:50:55,972 ERROR [main] - [🐳 [linagora/linshare-init:2.2]]- Could not start container
[55253ab273a9c6c7ee6f587b46ba8b82025df4ee] org.testcontainers.containers.ContainerLaunchException: Timed out waiting for log output matching '.*The following user provider for domain '.*' was successfully created.*'
...
[55253ab273a9c6c7ee6f587b46ba8b82025df4ee] [ERROR] Errors: 
[55253ab273a9c6c7ee6f587b46ba8b82025df4ee] [ERROR]   LinshareBlobExportMechanismTest.exportShouldFailWhenBlobDoesNotExist » NoClassDefFound
```
I think linshare-init got a problem in CI

File: server/data/data-api/src/main/java/org/apache/james/rrt/api/CyclicMappingException.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.rrt.api;
+
+public class CyclicMappingException extends RecipientRewriteTableException {
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/rrt/api/CyclicMappingException.java]: Cycle prevention is hard, especially for RRTs.

You only detect the easiest cases within this PR. The `A -> B + B -> A cases`. But there is some other way to be creating a RRT infinite loop that you are not preventing:

 - `A -> B + B -> C + C -> A` (adding intermediates) - these could still be detected upon insertion with Graph cycle algorithms
 - Mobilizing other RRT types like aliases, forwards and the like. Here the domain space is infinite, and you would need to check any RRT belonging to that domain.

My above point is just to highlight you provide only a partial security to the user - and that needs to be clear.

Finally, I want to bring in the **API consistency issues**. Why do you perform such checks on domains? And not on Aliases, forwards, and the like?

While I find it cool to be giving ideas on this concern, I'd like to underline that *this is a different issue* that needs additional discussions. And that should at the very least be separated from this PR.

File: server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/DomainAliasService.java [hunk 2]
```diff
@@ -81,6 +82,7 @@ public boolean hasAliases(Domain domain) throws DomainListException, RecipientRe
     }
 
     public void addDomainAlias(Domain sourceDomain, Domain destinationDomain) throws DomainListException, RecipientRewriteTableException {
+        checkCyclicMapping(sourceDomain, destinationDomain);
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/service/DomainAliasService.java]: IMO this should be a responsibility of the backend component and not of the backend layer.

**Comment by trantienduchn** [comment]
> As said above, the Cycle detection has nothing to do with "fix some bugs" and needs additional discussions.

Okay, fair, I will chop the last commit out of the PR

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[8116b0a73217c9b46d31c617a3223dd6bf736b54] [ERROR]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations:628 » ConditionTimeout
```

**Comment by Arsnael** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**